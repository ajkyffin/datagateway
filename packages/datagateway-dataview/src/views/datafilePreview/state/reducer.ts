import DATAFILE_PREVIEWER_DEFAULT from '../defaults';
import {
  DecrementDatafilePreviewerZoomLevelType,
  IncrementDatafilePreviewerZoomLevelType,
  ResetDatafilePreviewerZoomLevelType,
  ToggleDatafilePreviewerDetailsPanePayload,
  ToggleDatafilePreviewerDetailsPaneType,
} from './actions';
import { DGDataViewState } from '../../../state/app.types';

/**
 * Defines the shape of the redux state for {@link DatafilePreviewer}
 */
interface DatafilePreviewerState {
  isDetailsPaneShown: boolean;
  zoomLevel: number;
}

function toggleDatafilePreviewerDetailsPane(
  state: DGDataViewState,
  payload: ToggleDatafilePreviewerDetailsPanePayload
): DGDataViewState {
  return {
    ...state,
    isisDatafilePreviewer: {
      ...state.isisDatafilePreviewer,
      isDetailsPaneShown: payload.shouldShow,
    },
  };
}

function incDatafilePreviewerZoomLevel(
  state: DGDataViewState
): DGDataViewState {
  return {
    ...state,
    isisDatafilePreviewer: {
      ...state.isisDatafilePreviewer,
      zoomLevel:
        state.isisDatafilePreviewer.zoomLevel +
        DATAFILE_PREVIEWER_DEFAULT.zoomLevelStep,
    },
  };
}

function decDatafilePreviewerZoomLevel(
  state: DGDataViewState
): DGDataViewState {
  return {
    ...state,
    isisDatafilePreviewer: {
      ...state.isisDatafilePreviewer,
      // zoom level should not go lower than zoomLevelStep
      // otherwise will result in zoom level <= 0.
      zoomLevel: Math.max(
        DATAFILE_PREVIEWER_DEFAULT.minZoomLevel,
        state.isisDatafilePreviewer.zoomLevel -
          DATAFILE_PREVIEWER_DEFAULT.zoomLevelStep
      ),
    },
  };
}

function resetDatafilePreviewerZoomLevel(
  state: DGDataViewState
): DGDataViewState {
  return {
    ...state,
    isisDatafilePreviewer: {
      ...state.isisDatafilePreviewer,
      zoomLevel: DATAFILE_PREVIEWER_DEFAULT.zoomLevel,
    },
  };
}

const datafilePreviewerInitialState: DatafilePreviewerState = {
  isDetailsPaneShown: DATAFILE_PREVIEWER_DEFAULT.showDetailsPane,
  zoomLevel: DATAFILE_PREVIEWER_DEFAULT.zoomLevel,
};

const datafilePreviewerReducer = {
  [ToggleDatafilePreviewerDetailsPaneType]: toggleDatafilePreviewerDetailsPane,
  [IncrementDatafilePreviewerZoomLevelType]: incDatafilePreviewerZoomLevel,
  [DecrementDatafilePreviewerZoomLevelType]: decDatafilePreviewerZoomLevel,
  [ResetDatafilePreviewerZoomLevelType]: resetDatafilePreviewerZoomLevel,
};

export {
  datafilePreviewerInitialState,
  datafilePreviewerReducer,
  toggleDatafilePreviewerDetailsPane,
  incDatafilePreviewerZoomLevel,
  decDatafilePreviewerZoomLevel,
  resetDatafilePreviewerZoomLevel,
};
export type { DatafilePreviewerState };
