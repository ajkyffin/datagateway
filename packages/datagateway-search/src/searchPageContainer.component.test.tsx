import React from 'react';
import { ReactWrapper } from 'enzyme';
// history package is part of react-router, which we depend on
// eslint-disable-next-line import/no-extraneous-dependencies
import { createLocation } from 'history';

import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import { StateType } from './state/app.types';
import { initialState as dgSearchInitialState } from './state/reducers/dgsearch.reducer';
import { dGCommonInitialState } from 'datagateway-common';

import { createShallow } from '@material-ui/core/test-utils';
import { MemoryRouter } from 'react-router';
import SearchPageContainer from './searchPageContainer.component';
import { LinearProgress } from '@material-ui/core';

jest.mock('loglevel');

describe('SearchPageContainer - Tests', () => {
  let shallow;
  let state: StateType;

  const createWrapper = (path: string): ReactWrapper => {
    const mockStore = configureStore([thunk]);
    return shallow(
      <MemoryRouter initialEntries={[{ key: 'testKey', pathname: path }]}>
        <SearchPageContainer store={mockStore(state)} />
      </MemoryRouter>
    );
  };

  beforeEach(() => {
    shallow = createShallow({ untilSelector: 'Grid' });

    state = JSON.parse(
      JSON.stringify({
        dgcommon: dGCommonInitialState,
        dgsearch: dgSearchInitialState,
      })
    );
  });

  it('renders searchPageContainer correctly', () => {
    const wrapper = createWrapper('/');

    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly at /search/data route', () => {
    const wrapper = createWrapper('/search/data');

    expect(wrapper).toMatchSnapshot();
  });

  it('do not display loading bar loading false', () => {
    const wrapper = createWrapper('/search/data');

    expect(wrapper.exists(LinearProgress)).toBeFalsy();
  });

  it('display loading bar when loading true', () => {
    state = JSON.parse(
      JSON.stringify({
        dgcommon: { ...dGCommonInitialState, loading: true },
        dgsearch: {
          ...dgSearchInitialState,
          requestReceived: true,
        },

        router: {
          action: 'POP',
          location: createLocation('/'),
        },
      })
    );

    const wrapper = createWrapper('/search/data');

    expect(wrapper.exists(LinearProgress)).toBeTruthy();
  });
});
