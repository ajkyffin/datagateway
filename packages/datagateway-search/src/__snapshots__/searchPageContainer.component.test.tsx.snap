// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPageContainer - Tests renders correctly at /search/data route 1`] = `
<div>
  <WithStyles(ForwardRef(Grid))
    alignItems="flex-start"
    container={true}
    direction="column"
    justify="flex-start"
    spacing={2}
    style={
      Object {
        "margin": 0,
        "width": "100%",
      }
    }
  >
    <WithStyles(ForwardRef(Grid))
      id="container-search-filters"
      item={true}
    >
      <WithStyles(ForwardRef(Paper))
        style={
          Object {
            "height": "100%",
            "minWidth": 584,
            "width": "calc(70vw)",
          }
        }
      >
        <SearchBoxContainer
          initiateSearch={[Function]}
        />
      </WithStyles(ForwardRef(Paper))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</div>
`;

exports[`SearchPageContainer - Tests renders searchPageContainer correctly 1`] = `
<a
  href="/search/data"
  onClick={[Function]}
>
  Search data
</a>
`;
